### PR TITLE
Mason - what's old is new again!

### PR DESCRIPTION
**Pull request by cromedome**
Article detailing Mason support in Dancer2.

### CODE REVIEW

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+old - an L<HTML::Mason> plugin for Dancer2!
+
+=head2 But Santa, why? WHY?
+
+Mason is a solid templating engine, but it lacks a controller. A
+number of projects have built their own controllers (such as
+L<Request Tracker (RT)|https://bestpractical.com/request-tracker>), and
+others struggle without one. That's where Dancer2 comes in.
+
+L<Dancer2::Template::Mason> is a template engine for HTML::Mason. If
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:13]: Dancer2::T::M is more an interface to the engine than an engine itself. 

```suggestion
Well, uh oh no more!
L<Dancer2::Template::Mason> is a brand new interface  for HTML::Mason. If
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+HTML::Mason is your favorite template engine, you can now use it
+in your favorite web framework. If you have legacy Mason applications
+that would benefit from having a proper controller, or a general
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:16]: "proper" comes with a lot of judgement attached. :-) The original Mason handlers are stiffling and tightly coupled to Apache, but they are legit. 

```suggestion
that would benefit from having a Dancer-powered controller, or a general
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+refresh with some a more modern design, Dancer2 has your back.
+
+=head2 Does this mean I'm NOT on the Naughty List?
+
+Not at all! It's actually a pretty good gift, combining two powerful tools
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:21]: I read this as "No, no, you *totally* are on the naughty list, which I think is the opposite of what you're going for. :laughing:

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+that are already in the toolbox of a lot of Perl developers. Go ahead,
+you can thank Retro Santa now. I'll wait.
+
+HTML::Mason, like all frameworks, has its warts and annoyances, but also
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:25]: Like most of us, you have the natural tendency to apologize for the flaws of the thing before you (humbly) sing its praises. We're selling wares here. Sing loudly first, and maybe mention the flaws when it's necessary. Here, I think we can remove the paragraph altogether.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+a lot of upside, and not having HTML::Mason as part of the Dancer2
+ecosystem has been a glaring oversight. No more!
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:27]:
```suggestion

```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+=head2 How do I use it?
+
+First, make your layout (F<views/layouts/main.m>):
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:31]:
```suggestion
First, create your layout (F<views/layouts/main.m>):
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    $perl_version
+    $dancer_version
+    $settings
+    $request
+    $params
+    $vars
+    $session
+    </%args>
+
+That pile of variables in C<%args>? Like other template engines, Dancer2
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:56]:
```suggestion
That pile of variables in C<%args>? Like for its other template engines, Dancer2
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+gives you that automatically. HTML::Mason, unlike some other template
+engines (L<Template::Toolkit> for example) makes you explicitly declare
+variables before using them.
+
+Too many variables? You can always reference them via the C<%ARGS> hash.
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:61]:
```suggestion
gives you that automatically. HTML::Mason, unlike some other template
engines (L<Template::Toolkit> for example) prefers  you to explicitly declare
variables before using them. But, and don't tell anybody I told you this, if you don't wanna, you can reference them all via the C<%ARGS> hash too.
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    <p>Hello, <% $ARGS{ name } %>!</p>
+
+And finally, add a route that uses your Mason template:
+
+    get '/:name?' => sub {
+        my $name = body_paramaters->get( 'name' ) // 'Mystery Visitor';
+        template 'index', { name => $name };
+    };
+
+and Bob's your uncle! Now, when you go to C<http://localhost:5000/Uncle%20Bob>
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:74]:
```suggestion
and Bob's your uncle! Seriously! When you go to C<http://localhost:5000/Uncle%20Bob>
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    Hello, Uncle Bob!
+
+And if you go to C<http://localhost/:5000> you get:
+
+    Hello, Mystery Visitor!
+
+=head2 Care and feeding of your Mason templates
+
+Mason needs a little TLC to be at its best. To get the best performance
+out of Mason, you need to create a template cache in production, easily
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:86]:
```suggestion
Mason needs a little TLC to be at its best. To get the best performance, you need to create a template cache, easily
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+and Bob's your uncle! Now, when you go to C<http://localhost:5000/Uncle%20Bob>
+you'll see:
+
+    Hello, Uncle Bob!
+
+And if you go to C<http://localhost/:5000> you get:
+
+    Hello, Mystery Visitor!
+
+=head2 Care and feeding of your Mason templates
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:83]: This section has the problem that you apologize for both sides of the config. :-) You begin with "to get the best performance you have to do some work", and "to develop without going nutso you have to change the config". 

Turn that around! The default config is going to work beautifully with the dev process, and when the time comes to go to prod, you only have to change one line of config and VA-VA-VOOOOM!

Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:83]:
```suggestion

Off the shelf, the engine will re-compile the templates each time they are accessed, which is what one wants when developing. When the time comes for production and a little more ooomph is desired, the templates can be easily cached by adding the following to environments/production.yml:

template: "mason"
engines:
  template:
    mason:
      extension: m
      data_dir: "/path/to/your/app/var/"
      use_object_files: 1
      static_source: 1


```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+      template:
+        mason:
+          use_object_files: 0
+          static_source: 0
+
+=head2 Ok, where did this gift really come from? Retro Santa my %^$!!!
+
+Yanick is the brains and muscle behind this, with CromeDome doing what
+he does best (poking, prodding, agitating, and encouraging otherwise good
+people to do bad things).
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:117]:
```suggestion
There are actually two elves hiding behind the beard and Santa suit. Yanick reworked his original Dancer1 Mason plugin, thanks to CromeDome relentless poking, prodding, agitating, encouraging, testing, troubleshooting, and patch submitting.
```

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+=head2 Acknowledgements
+
+I'd like to thank Jonathan Swartz for the shout-out in the Mason
+documentation, and for contributing a Mason template engine for
```
Comment (yanick) [inline danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:122]:
```suggestion
documentation, and for contributing a Mason (v2) template engine for
```

**Review by yanick — APPROVED** [review]
Love it.

There are a lot of comments in there, but that's more because I'm a terrible interloper more than anything else. Take what makes sense, and if you have questions let's discuss!

:heart:

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 14]
```diff
@@ -0,0 +1,133 @@
+=head1 Dancer2::Plugin::Mason - What's old is new again!
+
+This year, Retro Santa paid us a visit and gave us a new gift of something
+old - an L<HTML::Mason> plugin for Dancer2!
+
+=head2 But Santa, why? WHY?
+
+Mason is a solid templating engine, but it lacks a controller. A
```
[the remark this answers is not in the document]
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Excellent :)

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+number of projects have built their own controllers (such as
+L<Request Tracker (RT)|https://bestpractical.com/request-tracker>), and
+others struggle without one. That's where Dancer2 comes in.
```
[the remark this answers is not in the document]
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Incorporated into the article.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 3 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:44]: Added... with a twist ;)

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 1 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Adapter was the word I was really looking for. But I had some momentum and ran with it anyhow.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 2 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Yes, yes it does. Not trying to start a flame war here.

Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: (to that end, I took the change, thanks!)

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 4 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Great point!

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 5 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Done.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 6 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Derp. Thanks, good catch. I think I started a different sentence and changed midstream.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 8 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Like how you turned that around!

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 11 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:102]: This makes perfect sense.

Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod:102]: This is great! Thank you.

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 12 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Took this, again with a twist ;)

File: danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod (hunk 13 above)
Comment (cromedome) [reply danceradvent/public/articles/2023/15-mason-whats-old-is-new-again.pod]: Done.

**Comment by cromedome** [comment]
@yanick, these were some great changes! I put a twist on a couple of them. Give this a quick look over when you can, and I'll merge after you :+1: it again.

**Review by yanick — APPROVED** [review]
I am Yanick and I approve of this messa-- advent entry. :-)

**Comment by cromedome** [comment]
Thanks! Merged!

**Pull request closed without merging**